### ISSUE TITLE
Disable current:site arg when in network mode

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#61]
Suggested in https://github.com/WordPoints/hooks-api/issues/59#issuecomment-167245342:

> As for the possibility of network-context events for which we'd want to register the current:site arg (and hence would want it to be a non-signature arg because we're in network context): Most network-wide events take place in the network admin or other network context when we're not actually in the context of a specific site. I think it would actually make sense to disable the current:site arg when in network-mode.

**Comment by JDGrimes** [comment]
To clarify, we're not talking about disabling the `site` entity, but the `current:site` hook arg. The difference is significant, because it means we're just in the context of the hooks API, not the entity API. (Though perhaps it is possible that we'll want to consider something like this in the entity API, too.) This is why we're talking about disabling this based on hooks network mode (which would be odd if we were talking about the entity itself—the hooks API shouldn't affect the behavior of the entity API!).

What #62 and #63 have brought to our attention is that the `site` entity isn't the only one that really can't exist as "current" in network-mode. In fact, all of the entities we currently have, except for the `user` entity, are in context of a specific site, so they can't be "current" except when there is a "current" site.

With that in mind, it seems we should explore making this a bit more generic, not just something we do with the current site arg.

Basically, what I have in mind is that we'd check if we were in network mode, and then check what the context of that particular type of entity which an arg is of is. If the entity can only exist in network context, then we'd want to disable the arg.

The main problem is then detecting whether the arg is a "current" arg. We can detect stateful args, but I'm not sure this is exactly the same thing as "current" args. The question is, is it possible to have a stateful arg that can be referenced outside of the context in which it exists?

What exactly is the goal of disabling these args in the first place? I suppose the idea is that disabling these args will have a cascading affect, and that any reactions that are fired which reference the disabled args would be disabled. I think that really that should be placed in its own ticket.

**Comment by JDGrimes** [comment]
> The main problem is then detecting whether the arg is a "current" arg. We can detect stateful args, but I'm not sure this is exactly the same thing as "current" args. The question is, is it possible to have a stateful arg that can be referenced outside of the context in which it exists?

This really depends. If we can set the context/GUID of an entity, then we are OK. Our concern is that when we grab the "current" context and we're out of context completely (or, really, just in the wrong context ID; a different context than what that entity is from). When we're supposed to be grabbing the "current" something, we're supposed to be grabbing it from the current context—which is impossible if that context isn't one in which such an entity directly exists.

So the answer to our question is that the concept of a "current" entity is a special case, and that in theory there is nothing which would prevent a stateful arg from existing which could supply an accurate GUID for its entity. The only potential roadblock that we need to overcome to make such a thing a reality is to make it possible to set the entity GUID in the API. Then any arg which can calculate the GUID can also set the entity properly, and does not need to be disabled.

In the case of "current" args, it just isn't possible for us to calculate the context (and therefore the GUID) and so the arg cannot work, because the entity value cannot really be supplied. This is not a problem with the API, but just the nature of what a "current" arg is. It is possible of course that there will also be those args that could in theory load the entity object with an accurate GUID, but are unable to do so for technical reasons.

In summary, not all stateful args necessarily have times when they need to be disabled. Just any arg that cannot calculate the context (and therefore the GUID) of the entity needs to be disabled at such a time.

We might well ask then, whether this is really limited to stateful args? It seems to me that there may well be times when a non-stateful arg is unable to calculate the GUID, and therefore would also be nonfunctional.

**Comment by JDGrimes** [comment]
From https://github.com/WordPoints/hooks-api/issues/75#issuecomment-169717104:

> An arg value might not be available just because it is only available conditionally. That is, no value is supplied by the action. The only difference between that and not being able to get a stateful arg's value because we aren't in a context in which it can (or should) exist, is that the arg value comes from a different place. Qualitatively, it is the same thing. 

That answers the last question we asked, as to whether this could apply to non-stateful args: yes.

Basically what we've learned from #74 and #75 is that we don't really need to take an active approach to "disabling" args. What we'll do is take a passive approach. So in generalizing this ticket beyond the `current:site` arg, all we are saying is that any stateful arg won't supply a value outside of the context in which the entity it represents exists.

The goal of this ticket then, is not to develop some active arg disablement API, but to make sure that we don't supply the value for the `current:site` arg outside of the context in which it should exist. This is important because the current site is set by WordPress even when we are really in network context (see #12). So that's why this arg needs special handling here, while others should just have this happen naturally. In other words, the site context needs to be restricted artificially by us, because WordPress runs certain things that are really in network context in the context of a particular site.

The approach that we'll take on this will probably involve the behavior of a context object and not this arg object specifically, depending on what course we take in #63.

**Comment by JDGrimes** [comment]
795d7a7 updates the entity context API so that `site` context is disabled when we are in the context of an Ajax request from the network admin screens (in addition to when we are in network context at other times).

When we retrieve an entity from the current state, we have to calculate its GUID based on its ID and context. When we are out of site context, the `site` context object will return `false` when the `get_current_id()` method is called. This will result in us coming up with a value of `false` for the entity context, which will mean our `get_the_GUID()` method will have to return `false`, too. What this means is that entities that only exist in site context will be disabled when in context of the network itself.

However, none of this actually fixes our issue here entirely. While we have, in essence, passive disablement of entities' `get_the_GUID()` methods, this doesn't stop an arg from being supplied—it only means that we will be unable to get a GUID for it.

Like we said above, we need to actually apply active disablement directly to this arg, since a value can be supplied for it in WordPress when it really shouldn't be (when really in network context). The solution to this will be to use the `wordpoints_is_network_context()` function introduced in 795d7a7 in the current site arg class to return `false` when in network context.

The only questions would be whether we should make this more generic, and whether the entity should be returned at all when no value is set.

**Comment by JDGrimes** [comment]
The first thing to consider is whether this should be generalized into something handled in the main arg class. I guess we really can't, because we can't know for sure what context the entity might correspond to (the `site` entity has a corresponding `site` entity context, but there is no guarantee that it will always be that way). I suppose that we could let the arg specify the necessary context in a property value, but I think that it is overkill to generalize this at the moment, when we really don't know of any other case where an entity value is available from the state when it should be out of context.

The main question that we need to answer then, is what we should do in the case when an arg cannot retrieve its value. Currently there is no special handling for this, so the entity object is still returned, just without the value set (or with it set to `false`).

We've actually already explored that in #75, and concluded that we should just keep doing what we're doing, at least for now. I do think that we should skip trying to set the entity value if the arg value is falsey, since that doesn't make much sense, and might even lead to unexpected behavior.
